skeleton/src/main/java/com/dropbox/kaiken/skeleton/skeleton/core/KaikenUserServicesProvider.kt

```diff
@@ -0,0 +1,67 @@
+package com.dropbox.kaiken.skeleton.skeleton.core
+
+import com.dropbox.kaiken.scoping.AppServices
+import com.dropbox.kaiken.scoping.TeardownHelper
+import com.dropbox.kaiken.scoping.UserServices
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+
+/**
+ * Skeleton implementation of [UserServicesProvider] that registers itself with a [UserManager].
+ *
+ */
+class KaikenUserServicesProvider
+constructor(
+        private val applicationServices: AppServices,
+        private val userServicesFactory: (AppServices, SkeletonUser) -> UserServices,
+) : SkeletonUserServicesProvider {
+    private val userServicesMap = mutableMapOf<String, KaikenUserServices>()
+
+    init {
+        //TODO Mike figure out if we want another scope to launch from
+        GlobalScope.launch {
+            (applicationServices as KaikenAppServices).userManager().getUserState().collect { userState ->
+                userState.usersRemoved.forEach { user ->
+                    teardownUserServicesOf(user.userId)
+                }
+                //TODO Mike figure out what we want to do for the base user type
+                userState.usersAdded.forEach { user ->
+                    initUserServicesOf(
+                            SkeletonUser(
+                                    user.userId,
+                                    SkeletonOauth2(user.accessToken),
+                                    user.isActiveUser
+                            )
+                    )
+                }
+            }
+        }
+    }
+
+    override fun initUserServicesOf(user: SkeletonUser) {
+        synchronized(userServicesMap) {
+            val userServices: KaikenUserServices = userServicesFactory(applicationServices, user) as KaikenUserServices
+            userServicesMap[user.userId] = userServices
+        }
+    }
+
+    override fun teardownUserServicesOf(userId: String) {
+        synchronized(userServicesMap) {
+            val teardownHelper = userServicesMap.remove(userId)?.getTeardownHelper()
```

Comment:
I think we can simplify this as 
```
userServicesMap.remove(userId)?.getTeardownHelper()?.teardown()
```

and remove `dispatchTeardown` 

Also, what thread should the `teardown` occur on? I assume the main thread?

Reply:
Changed both init and teardown to be `suspend` functions and added a context switch to main dispatcher. I would assume creating user services (dagger components) off main thread would be weird too

Reply:
thought about this some more, if teardown is main and creating is on main, I'm going to instead just launch in main scope. There is minimal work done here I'm ok blocking.